conn.go

```diff
@@ -76,26 +79,109 @@ func (r *filteredConn) ReadFrom(b []byte) (n int, addr net.Addr, err error) {
 	select {
 	case <-timeout:
 		return 0, nil, errTimeout
-	case pkt := <-r.recvBuffer:
-		n := pkt.n
-		err := pkt.err
+	case msg := <-r.recvBuffer:
+		n := msg.N
+		err := msg.Err
 		if l := len(b); l < n {
 			n = l
 			if err == nil {
 				err = io.ErrShortBuffer
 			}
 		}
-		copy(b, pkt.buf[:n])
-		r.source.bufPool.Put(pkt.buf[:r.source.packetSize])
-		if pkt.oobBuf != nil {
-			r.source.bufPool.Put(pkt.oobBuf[:r.source.packetSize])
-		}
-		return n, pkt.addr, err
+		copy(b, msg.Buffers[0][:n])
+
+		r.source.returnBuffers(msg.Message)
+
+		return n, msg.Addr, err
 	case <-r.closed:
 		return 0, nil, errClosed
 	}
 }
 
+func (r *filteredConn) ReadBatch(ms []ipv4.Message, flags int) (int, error) {
+	if flags != 0 {
+		return 0, errNotSupported
+	}
+
+	if len(ms) == 0 {
+		return 0, nil
+	}
+
+	var timeout <-chan time.Time
+
+	if deadline, ok := r.deadline.Load().(time.Time); ok && !deadline.IsZero() {
+		timer := time.NewTimer(deadline.Sub(time.Now()))
+		timeout = timer.C
+		defer timer.Stop()
+	}
+
+	var msgs []messageWithError
+
+	defer func() {
+		for _, msg := range msgs {
+			r.source.returnBuffers(msg.Message)
+		}
+	}()
+
+	// We must read at least one message.
+	select {
+	//goland:noinspection GoNilness
+	case <-timeout:
+		return 0, errTimeout
+	case msg := <-r.recvBuffer:
+		msgs = append(msgs, msg)
+		if msg.Err != nil {
+			return 0, msg.Err
+		}
+	case <-r.closed:
+		return 0, errClosed
+	}
+
+	// After that, it's best effort. If there are messages, we read them.
+	// If not, we break out and return what we got.
```

Comment:
In theory there are `syscall.MSG_WAITONE/WAITALL` flags, but that's unix only, and I am not bothered to add support for these.

The usecase here is quic, and that doesn't use any flags.